src/ring_theory/ideals.lean

```diff
@@ -364,3 +306,168 @@ instance : is_ring_hom (lift S f H) :=
+end
+
+class is_local_ring (α : Type u) [comm_ring α] : Prop :=
+(zero_ne_one : (0:α) ≠ 1)
+(is_local : ∀ (a : α), (is_unit a) ∨ (is_unit (1 - a)))
+
+namespace is_local_ring
+variable [is_local_ring α]
+
+instance : zero_ne_one_class α :=
+{ ..‹comm_ring α›, ..‹is_local_ring α› }
+
+lemma is_unit_or_is_unit_one_sub_self (a : α) :
+  (is_unit a) ∨ (is_unit (1 - a)) :=
+is_local a
+
+lemma is_unit_of_mem_nonunits_one_sub_self (a : α) (h : (1 - a) ∈ nonunits α) :
+  is_unit a :=
+or_iff_not_imp_right.1 (is_local a) h
+
+lemma is_unit_one_sub_self_of_mem_nonunits (a : α) (h : a ∈ nonunits α) :
+  is_unit (1 - a) :=
+or_iff_not_imp_left.1 (is_local a) h
+
+lemma nonunits_add {x y} (hx : x ∈ nonunits α) (hy : y ∈ nonunits α) :
+  x + y ∈ nonunits α :=
+begin
+  rintros ⟨u, hu⟩,
+  apply hy,
+  suffices : is_unit ((↑u⁻¹ : α) * y),
+  { rcases this with ⟨s, hs⟩,
+    use u * s,
+    convert congr_arg (λ z, (u : α) * z) hs,
+    rw ← mul_assoc, simp },
+  rw show (↑u⁻¹ * y) = (1 - ↑u⁻¹ * x),
+  { rw eq_sub_iff_add_eq,
+    replace hu := congr_arg (λ z, (↑u⁻¹ : α) * z) hu,
+    simpa [mul_add] using hu },
+  apply is_unit_one_sub_self_of_mem_nonunits,
+  exact mul_mem_nonunits_right hx
+end
+
+variable (α)
+
+def nonunits_ideal : ideal α :=
+{ carrier := nonunits α,
+  zero := zero_mem_nonunits.2 $ zero_ne_one α,
+  add := λ x y hx hy, nonunits_add hx hy,
+  smul := λ a x, mul_mem_nonunits_right }
+
+instance nonunits_ideal.is_maximal : (nonunits_ideal α).is_maximal :=
+begin
+  rw ideal.is_maximal_iff,
+  split,
+  { intro h, apply h, exact is_unit_one },
+  { intros I x hI hx H,
+    erw not_not at hx,
+    rcases hx with ⟨u,rfl⟩,
+    simpa using I.smul_mem ↑u⁻¹ H }
+end
+
+lemma max_ideal_unique :
+  ∃! I : ideal α, I.is_maximal :=
+⟨nonunits_ideal α, nonunits_ideal.is_maximal α,
+  λ I hI, hI.eq_of_le (nonunits_ideal.is_maximal α).1 $
+  λ x hx, hI.1 ∘ I.eq_top_of_is_unit_mem hx⟩
+
+variable {α}
+
+@[simp] lemma mem_nonunits_ideal (x) :
+  x ∈ nonunits_ideal α ↔ x ∈ nonunits α := iff.rfl
+
+end is_local_ring
+
+lemma local_of_nonunits_add (hnze : (0:α) ≠ 1)
+  (h : ∀ x y ∈ nonunits α, x + y ∈ nonunits α) : is_local_ring α :=
+{ zero_ne_one := hnze,
+  is_local := λ x, or_iff_not_imp_left.mpr $ λ hx,
+  begin
+    by_contra H,
+    apply h _ _ hx H,
+    simp [-sub_eq_add_neg, add_sub_cancel'_right]
+  end }
+
+lemma local_of_unique_max_ideal (h : ∃! I : ideal α, I.is_maximal) : is_local_ring α :=
+let ⟨I, h₁, h₂⟩ := h in
+let hnze : (0:α) ≠ 1 := (λ H, h₁.1 $ I.eq_top_iff_one.2 $ H ▸ I.zero_mem) in
+local_of_nonunits_add hnze $ λ x y hx hy H,
+let ⟨Ix, Ixmax, Hx⟩ := exists_max_ideal_of_mem_nonunits hx in
+let ⟨Iy, Iymax, Hy⟩ := exists_max_ideal_of_mem_nonunits hy in
+have xmemI : x ∈ I, from ((h₂ Ix Ixmax) ▸ Hx),
+have ymemI : y ∈ I, from ((h₂ Iy Iymax) ▸ Hy),
+h₁.1 $ I.eq_top_of_is_unit_mem (I.add_mem xmemI ymemI) H
+
+class is_local_ring_hom [comm_ring β] (f : α → β) [is_ring_hom f]
```

Comment:
Can this just `extend is_ring_hom`?

Reply:
I considered doing that. But then it becomes harder to state the condition that a particular ring hom should be a local ring hom. E.g. for locally ringed spaces: we'll have some `stalk_map x` that is automatically a ring hom. And now we want to add the condition that for all `x` this is in fact a local ring hom. Hence I thought I would structure it the way I did.
But if we can somehow make this work while using `extends` that would be even better.

Reply:
Hmmm, I realise this is nonsense. We can have both.
So now `is_local_ring_hom` extends `is_ring_hom`.